### PR TITLE
Add `DensePaddingDataLoader` for flexible (padded) dense batching of `BaseData` objects

### PR DESCRIPTION
**Pull request by amorehead**
### 🚀 The new functionality, motivation and pitch

* This PR adds a new `DensePaddingDataLoader` class that enables flexible dense batching of `BaseData` objects.
* Such functionality supports fast and direct conversion of PyG `Data` or `HeteroData` objects into CV/NLP-driven (dense-batching) model training pipelines.

### Alternatives

No alternatives exist for this functionality, including PyG's `DenseDataLoader` which assumes each (sub)graph's node count must be identical to all other (sub)graphs.

### Additional context

This new data loader enables one to easily experiment with message passing algorithms that operate on dense tensor representations of batched graph objects without having to redesign one's PyG-driven `Dataset` class. This idea was initiated per https://github.com/pyg-team/pytorch_geometric/discussions/8516 and raised as an issue in https://github.com/pyg-team/pytorch_geometric/issues/8517.

### CODE REVIEW

File: torch_geometric/loader/dense_padding_data_loader.py [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
+        exclude_keys=exclude_keys,
+    )
+
+    batch._num_graphs = len(data_list)
+    batch.mask_dict = mask_dict  # NOTE: `mask_dict` must be a public attribute of `Batch` for auto-device onloading to work.
+
+    return batch
+
+
+class DensePaddingCollater:
```
[the remark this answers is not in the document]
Comment (amorehead) [reply torch_geometric/loader/dense_padding_data_loader.py:324]: Hi, @rusty1s. It's possible we may be able to reuse existing assets. I'm not too familiar with the `PadTransform` utility. Does it already support padding both float and non-float (e.g., int, bool) tensors?

[the remark this answers is not in the document]
Comment (rusty1s) [reply torch_geometric/loader/dense_padding_data_loader.py:324]: Yes, that should be supported.

**Comment by Mjvolk3** [comment]
Any word on this?

**Comment by amorehead** [comment]
Hey, @Mjvolk3. I've been too busy to update this PR sadly. If you'd like to take the lead to finish it, feel free to go for it!

**Comment by Mjvolk3** [comment]
@amorehead Could you give me a recap on it's status. I copied the code into my own repo to give it a go. I see you get a mask and the padded output which I expected. I thought that that the `edge_index` would be `adj` matrix (BxNxN dense version of edge index.). Is this the expected behavior? Thanks.

**Comment by Mjvolk3** [comment]
I see now that we would need an equivalent `ToDense` but for HeteroData to get the dense adjacency matrix.

**Comment by Mjvolk3** [comment]
@amorehead I've started using this code in my own repo and it seems to be working ok. Do you remember if there is anything that needs to be done other than satisfying the Pep8 reqs?

**Comment by amorehead** [comment]
I think that should be everything!
